When writing LLVM IR output demangled fn name in comments

stepancheg opened the pull request:
`--emit=llvm-ir` looks like this now:

```
; <alloc::vec::Vec<T> as core::ops::index::IndexMut<core::ops::range::RangeFull>>::index_mut
; Function Attrs: inlinehint uwtable
define internal { i8*, i64 } @"_ZN106_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..index..IndexMut$LT$core..ops..range..RangeFull$GT$$GT$9index_mut17h7f7b576609f30262E"(%"alloc::vec::Vec<u8>"* dereferenceable(24)) unnamed_addr #0 {
start:
  ...
```

cc https://github.com/integer32llc/rust-playground/issues/15

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

stepancheg commented:
Updated PR with `call` instruction comments:

```
; std::collections::hash::table::calculate_offsets
; Function Attrs: inlinehint uwtable
define internal void @_ZN3std11collections4hash5table17calculate_offsets17hccd439e3af053e99E({ i64, [0 x i8], i64, [0 x i8], i8, [7 x i8] }* noalias nocapture sret dereferenceable(24), i64, i64, i64) unnamed_addr #0 {
start:
  %tmp_ret = alloca { i64, [0 x i8], i8, [7 x i8] }
; call std::collections::hash::table::round_up_to_next
  %4 = call i64 @_ZN3std11collections4hash5table16round_up_to_next17hb5dd1061b779aff5E(i64 %1, i64 %3)
  ...
```

stepancheg commented:
Sample source code and output: [gist](https://gist.github.com/stepancheg/33edd364f23826a948023d57954e2f4e).

nagisa commented on src/libstd/lib.rs:
```diff
@@ -471,6 +471,13 @@ pub mod time;
+pub mod demangle_for_rustc {
+    pub use sys_common::backtrace::demangle;
+    pub use sys_common::backtrace::PrintFormat;
+}
```
IMO it is a shame that we’re exposing more unstable stuff from libstd. Would it be possible to extract the functionality into a crate and make both libstd and librustc_trans use that?

stepancheg replied:
Yes, I'll try to do that now. What should be crate name, `librust_demange`?

nagisa replied:
Well there’s technically a [rustc-demangle](https://crates.io/crates/rustc-demangle) already, although I do not know how easy/hard it is to integrate an external crate into libstd. I know there are a few already integrated in other places such as librustc_trans, could take a look there.

stepancheg replied:
@alexcrichton any suggestions on that issue?

nagisa commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -503,8 +504,133 @@ LLVMRustWriteOutputFile(LLVMTargetMachineRef Target, LLVMPassManagerRef PMR,
+  // return empty string if demangle failed
+  // or if name does not need to be demangled
+  StringRef CallDemangle(StringRef name) {
+    if (!Demangle) {
```
Could the constructor use a default no-op demangle callback instead?

stepancheg replied:
I can just remove the comment. No-op demangle would be overengineering IMHO.

nagisa commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -503,8 +504,133 @@ LLVMRustWriteOutputFile(LLVMTargetMachineRef Target, LLVMPassManagerRef PMR,
+class RustAssemblyAnnotationWriter : public AssemblyAnnotationWriter {
+  DemangleFn Demangle;
+  static const size_t BufCap = 500;
+  char Buf[BufCap];
```
Could the output buffer size derived from the input symbol instead? Something like a std::vector that only ever expands when it sees an input symbol longer than its current output buffer.

stepancheg replied:
OK, I'll do that.

nagisa commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -503,8 +504,133 @@ LLVMRustWriteOutputFile(LLVMTargetMachineRef Target, LLVMPassManagerRef PMR,
+                            formatted_raw_ostream &OS) override {
+    const char *Name;
+    const Value *Value;
+    if (const CallInst *CI = dyn_cast<CallInst>(I)) {
```
Some other cases where a demangle could be relevant is taking the address of the function.

That would be `store %place, @function`.

Maybe it is worth to always look at the instruction operands and output a demangled version if there’s a function operand?

stepancheg replied:
I'll keep printing only `invoke` and `call` for now (which covers 99% of cases), and probably do more printing in another PR, if that's OK with you.

nagisa replied:
Yeah, that’s fine.

nagisa commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -503,8 +504,133 @@ LLVMRustWriteOutputFile(LLVMTargetMachineRef Target, LLVMPassManagerRef PMR,
+  bool runOnModule(Module &M) override {
+    RustAssemblyAnnotationWriter AW(Demangle);
+
+    // Copy-paste of PrintModulePass::run
```
Remove the comments?

stepancheg replied:
OK.

nagisa commented:
r? @nagisa

stepancheg commented:
Updated the PR.

Removed comments, replace fixed size buf with `std::vector`.

`librustc_trans` now uses `rustc-demangle` from crates, but `std` still contains private implementation of `demangle` (because I don't know what could happen if std will depend on crates from cargo, removal of `demangle` from std could be done later).

nagisa commented:
Awesome, thanks!

@bors r+

bors commented:
:pushpin: Commit 8f6ac5b has been approved by `nagisa`

<!-- @bors r=nagisa 8f6ac5b0bec357ec6ef7c68a01a7b7e8397a3f57 -->

stepancheg commented:
After changing to `rustc-demangle`, I overlooked that comments now contain function hashes, which they shouldn't.

I've pushed a fix to the branch. The only change is this:

```
diff --git a/src/librustc_trans/back/write.rs b/src/librustc_trans/back/write.rs
index cfb5045afe..562d717115 100644
--- a/src/librustc_trans/back/write.rs
+++ b/src/librustc_trans/back/write.rs
@@ -537,7 +537,7 @@ unsafe fn optimize_and_codegen(cgcx: &CodegenContext,
                     Err(_) => return 0,
                 };

-                if let Err(_) = write!(cursor, "{}", demangled) {
+                if let Err(_) = write!(cursor, "{:#}", demangled) {
                     // Possible only if provided buffer is not big enough
                     return 0;
                 }
```

Someone please do `r+` again please.

stepancheg commented:
@bors r-

(Can I do that?)

bors commented:
@stepancheg: :key: Insufficient privileges: Not in reviewers

Mark-Simulacrum commented:
@bors r=nagisa

bors commented:
:pushpin: Commit b62bdaa has been approved by `nagisa`

<!-- @bors r=nagisa b62bdaafe038b8933fac5df5fa0fa5ddbaf176b7 -->

bors commented:
:hourglass: Testing commit b62bdaafe038b8933fac5df5fa0fa5ddbaf176b7 with merge d41b791c1af56b1f538cc6f77e597b7ad5b4d223...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3764), [status-travis](https://travis-ci.org/rust-lang/rust/builds/249033715?utm_source=github_status&utm_medium=notification)
Approved by: nagisa
Pushing d41b791c1af56b1f538cc6f77e597b7ad5b4d223 to master...
